### ISSUE TITLE
Opaque with_scope behaviour

### ISSUE THREAD

**Issue by sl0thentr0py** [getsentry/sentry-ruby#1855]
Creating this issue for further discussion.

If you do the following, the exception will be captured but the tag won't be used (because that scope is thrown away in [an ensure](https://github.com/getsentry/sentry-ruby/blob/332315907185e6ffe05aa64dc751cdd8ab9f5448/sentry-ruby/lib/sentry/hub.rb#L60-L61)).
```ruby
    Sentry.with_scope do |scope|
      scope.set_tags(my_tag: "my value")
      1 / 0
    end
```

If you do the following, the exception will be captured _and_ the tag will be visible.
```ruby
    Sentry.with_scope do |scope|
      scope.set_tags(my_tag: "my value")

      begin
        1 / 0
      rescue => e
        Sentry.capture_exception(e)
      end
    end
```

_Originally posted by @sl0thentr0py in https://github.com/getsentry/sentry-docs/pull/5328#issuecomment-1194148134_

**Comment by st0012** [comment]
I think this is counter-intuitive and probably should be considered a bug.

**Comment by st0012** [comment]
After testing this, I think the actual problem is: should `with_scope` capture exceptions? 

If it should, then we can

1. report the exception with the inner scope's data
1. pop the scope
1. re-raise the exception
1. the rest of the program continues with the original scope

(Because we now flag captured exceptions, it won't cause duplicated reporting)

If it shouldn't, we need to **not** popping the scope so other integrations (e.g. Rack middleware) can capture it with the inner scope's data. However, this can lead to other issues like polluting the request transaction

Ofc, we can also just keep the current behavior.

**Comment by sl0thentr0py** [comment]
@st0012 so there are some things to consider here.
1. `with_scope` belongs to the [Unified API](https://develop.sentry.dev/sdk/unified-api/) so we can't just change it outright, even though I personally don't like these semantics.
2. we could introduce a new API that does the better semantics (capture/pop/re-raise) as you said, maybe call it `try_with_scope` (name could be better potentially)
3. if people like that more, we could deprecate `with_scope` across SDKs and then eventually remove it in some major.

**Comment by st0012** [comment]
Don't other SDKs have the same issue? What are their solutions to this?

**Comment by sl0thentr0py** [comment]
Yes the issue is the same across SDKs, so we can drive the change process by doing it in ruby first. But we cannot change the old `with_scope`, so it will have to be a new API.

**Comment by st0012** [comment]
That's fair. I'm not rushing to change it to be clear.
Will adding a keyword argument to `with_scope` be an option? Like `Sentry.with_scope(capture: true) { }`

**Comment by sl0thentr0py** [comment]
hmm yea good point! 
As long as the old behavior is preserved/backward compatible, we should be good.

**Comment by benoittgt** [comment]
I am interested to work on this one and add the option to "capture". But I think it should be the default on a major version.
